File: go/tree/102_level_order.go
package tree

// BFS levelOrder 层级遍历，queue
func levelOrder(root *TreeNode) [][]int {

	if root == nil {
		return nil
	}

	ans := make([][]int, 0)

	queue := make([]*TreeNode, 0)
	queue = append(queue, root)

	for len(queue) > 0 {
		// 当前层的节点个数
		length := len(queue)
		level := make([]int, 0)
		// also can use for range instead
		for i := 0; i < length; i++ {
			head := queue[0]
			level = append(level, head.Val)
			queue = queue[1:]

			if head.Left != nil {
				queue = append(queue, head.Left)
			}
			if head.Right != nil {
				queue = append(queue, head.Right)
			}
		}

		ans = append(ans, level)
	}

	return ans

}

File: rust/src/tree/_102_level_order.rs
use super::TreeNode;
use std::cell::RefCell;
use std::rc::Rc;

// level_order 层序遍历 BFS
pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
    // if root == None {
    //     return vec![]
    // }

    if root.is_none() {
        return vec![];
    }

    use std::collections::VecDeque;

    let mut ans = vec![];

    // 因为需要在头部进行删除元素，VecDeque 比Vec 更合适
    let mut queues = VecDeque::new();
    queues.push_back(root);

    while !queues.is_empty() {
        let mut level_res: Vec<i32> = vec![];

        for _ in 0..queues.len() {
            let head = queues.pop_front().unwrap().unwrap();

            level_res.push(head.borrow().val);

            if head.borrow().left.is_some() {
                queues.push_back(head.borrow_mut().left.take());
            }

            if head.borrow().right.is_some() {
                queues.push_back(head.borrow_mut().right.take());
            }
        }

        ans.push(level_res);
    }

    ans
}
